yamlpatch/patch.go

```diff
@@ -0,0 +1,255 @@
+// Copyright (c) 2021 Palantir Technologies. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package yamlpatch is a library for applying RFC6902 JSON patches to yaml documents.
+// It leverages go-yaml v3's Node type to preserve comments, ordering, and most formatting.
+package yamlpatch
+
+import (
+	"bytes"
+	"encoding/json"
+	"reflect"
+
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v3"
+)
+
+var (
+	// IndentSpaces configures the yaml encoder used to marshal patched objects.
+	IndentSpaces = 2
```

Comment:
Is this variable intentionally exported so that consumers can modify it? If so, does it make sense to be global for the package, or to have this be something that is configured per operation?

Reply:
Yeah, admittedly this was me being a little lazy with how to configure this. I've just pushed a commit which unexports this variable in favor of an additional `ApplyIndent` entrypoint which allows specifying the indent length.

Reply:
Thanks!

Mostly on-board, but just wanted to discuss how to handle parameters in general before merging this since, while playing around with this locally, there was another thing I wanted to configure.

Right now, the value of the Patch is converted to a YAML node by using `json.Marshal`. However, this isn't ideal for Go structs that define YAML marshalling -- behavior like `omitempty`, custom field names and field ordering are all not respected. Since this is likely to be used by Go code that operates on YAML structs, I think that wanting to marshal while respecting this is a common use case. Although performing a YAML struct -> YAML bytes -> JSON struct translation on the caller side solves a lot of the issues, in this case you lose field ordering.

This problem can be solved by using `yaml.Marshal` rather than `json.Marshal` in `valueToYAMLNode`, and I think that this is something that is reasonable to be able to configure (either as "YAML mode" or more generally by allowing callers to provide an `interface{} -> []byte` function).

Given that we now have scenarios where we may want to configure indentation and marshal mode and there may be more in the future, wondering if we should go an OO route where we configure a marshaler and have `Apply` be a function on the marshaler, or have `Apply` accept `Param` objects and define these configurable things that way.